internal/consensus/state.go

```diff
@@ -1377,27 +1389,37 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	pbtsEnabled := cs.isPBTSEnabled(height)
+	if pbtsEnabled && !cs.Proposal.Timestamp.Equal(cs.ProposalBlock.Header.Time) {
+		logger.Debug("prevote step: proposal timestamp does not match block time; prevoting nil")
+		cs.signAddVote(types.PrevoteType, nil, types.PartSetHeader{}, nil)
+		return
+	}
+
 	// Timestamp validation using Proposed-Based TimeStamp (PBTS) algorithm.
 	// See: https://github.com/cometbft/cometbft/blob/main/spec/consensus/proposer-based-timestamp/README.md
-	if cs.isPBTSEnabled(height) {
-		if !cs.Proposal.Timestamp.Equal(cs.ProposalBlock.Header.Time) {
-			logger.Debug("prevote step: proposal timestamp not equal; prevoting nil")
-			cs.signAddVote(types.PrevoteType, nil, types.PartSetHeader{}, nil)
-			return
-		}
-
-		if cs.Proposal.POLRound == -1 && cs.LockedRound == -1 && !cs.proposalIsTimely() {
-			logger.Debug("prevote step: Proposal is not timely; prevoting nil",
-				"proposed",
-				cmttime.Canonical(cs.Proposal.Timestamp).Format(time.RFC3339Nano),
-				"received",
-				cmttime.Canonical(cs.ProposalReceiveTime).Format(time.RFC3339Nano),
-				"msg_delay",
-				cs.state.ConsensusParams.Synchrony.MessageDelay,
-				"precision",
-				cs.state.ConsensusParams.Synchrony.Precision)
+	if pbtsEnabled && cs.Proposal.POLRound == -1 {
+		isProposalTimely := cs.proposalIsTimely()
+
+		// For logging and metrics
+		lowerBound, upperBound := cs.timelyProposalMargins()
+		timestampDifference := cs.ProposalReceiveTime.Sub(cs.Proposal.Timestamp)
+		cs.metrics.ProposalTimestampDifference.With(
```

Comment:
We should produce this metric, and maybe print (in debug mode) this difference even when PBTS is not in use.

This does not make sense, for PBTS, when we have `PolRound >= 0`.

See: https://github.com/cometbft/cometbft/issues/2323#issuecomment-1966601461